### ISSUE TITLE
Upgrader - remnants of old themes around, prevents package install

### ISSUE THREAD

**Issue by sbulen** [SimpleMachines/SMF#3962]
Just tried testing an install of a package, and received the following error:  
![image](https://cloud.githubusercontent.com/assets/23568484/23839990/00fa3d14-075f-11e7-8a47-e7aed5d554ec.png)

It appears there are references to old themes in the themes table.  The package cannot be installed because it cannot make a backup first.  It can't make the backup, because the core theme directory doesn't exist in 2.1.

I've confirmed the old entries are in the themes table...  At this point, I am assuming that's where the reference to core comes from in the above message.

**Comment by sbulen** [comment]
Confirmed, backups first try to backup your themes...  I believe the upgrader should clear out all the old theme data.  This is the applicable code in subs-package.php:
![image](https://cloud.githubusercontent.com/assets/23568484/23873724/3a2a1f0a-07f0-11e7-9aa4-f822cd6fe5c2.png)

**Comment by sbulen** [comment]
Note that the Upgrader already attempts to clean these up, but I see issues with its current approach: 
 - It only cleans up core
 - It only cleans up core if it still exists on the file system
 - It assumes all other theme data should be left alone in the themes table

I believe we are now working under the premise that the themes are going to be wiped out as part of the upgrade.  2.0 themes are not guaranteed to work in 2.1.  We will ask folks to get rid of their old Themes folder as part of the upgrade process.  Thus, the old themes, including core, should _not_ be found in the file system.  

So...  

I think that the upgrader logic should be changed to wipe out all details for _all_ themes except the new default theme.  This cleanup should not be dependent on what it finds in the file system.  

Feedback welcome.

**Comment by illori** [comment]
2.0 themes will never work on SMF 2.1. but i have yet to see it advised to user to delete their old themes at the time of the upgrade. also some may not delete core as it is a theme they did not install in the first place if they are told to delete themes.

**Comment by sbulen** [comment]
@illori :  We need a common vision for 2.0 => 2.1 migration, in order to know what to build to.  

The only option that I see that makes sense is the one I have been told from the beginning, when I asked how to help out on the beta:  Delete all folders except attachments, avatars & settings.php.  Copy new files over & run the upgrader.

That is the only process supported by 2.1 today that will work cleanly on a 2.0 => 2.1 upgrade.  Any other option requires new design & a further phase of development & rework.  And LOTS more testing...  I'm not sure that makes sense at all given where we are at.   Following the process above, you're actually pretty close today.   If you want a significant change in approach, that is as-yet not even thought through, well, we will continue to see moving goalposts and unnecessarily extending the beta.  

I strongly advise the 'clean break' approach spelled out above, and telling folks to plan on re-applying themes & mods in 2.1.  It's time to drive towards closure.  

I believe any other decision will add another 1 year+ to the beta.   

We already know 2.0 themes & mods don't work on 2.1...  We should act accordingly, and work on an upgrade process that provides the cleanest user experience.

**Comment by illori** [comment]
http://wiki.simplemachines.org/smf/Upgrading is what we officially tell users on how to upgrade. your method is not what we tell users. no where there does it say to delete folders anywhere.

**Comment by sbulen** [comment]
@illori - On the SMF downloads page, the upgrader says it is going to automatically remove all mods.  In fact, there is existing code in the 2.0 upgrader to remove all themes and mods, but it is commented out due to issues.   And it was probably not needed in a 2.0.x =>2.0.x upgrade anyway, so it was not a priority to fix it.

The question is whether that needs to be fixed & restored (in one form or another) for the 2.0 => 2.1 upgrade.  I think we are asking for trouble if we don't, due to the significant differences between 2.0 & 2.1.  

I.e., the reason we don't tell them to clear it out is that we had code to do that as part of the upgrade.

**Comment by illori** [comment]
in the upgrade to SMF 2.1 all hooks are removed and the files are overwritten so any calls to other files are removed. you dont need to remove folders to get the effect of uninstalling mods.

**Comment by sbulen** [comment]
Understood.

I started off in late December trying to do an upgrade.  The upgrade instructions say to copy the files to a "new folder".  I was unsure whether that meant a clean directory or just not to do it directly in production.  I posted this online to try to get clarification:
http://www.simplemachines.org/community/index.php?topic=550513.0

The answer was: Different folks were doing it different ways.  

I am now coming to the realization that both methods (overwrite & clear it out first) probably need to be supported.  The 'overwrite' method because that seems to be the way the 2.1 install is headed - it's simpler for users to accomplish.  The 'clear it out' method because the support site has used that method as a last resort to fix sites with botched installs/deinstalls.

OK.  I'll modify the PR to find another way to fix the backup issue.  

(The Sheldon Cooper in me still wishes to delete those old files...)
